certification/pyxis/builder_test.go

```diff
@@ -39,4 +43,218 @@ var _ = Describe("Pyxis Builder tests", func() {
+		f := filepath.Join(tmpdir, "does-not-exist")
+		It("should fail when the file does not exist", func() {
+			err := readAndUnmarshal(f, nil)
+			Expect(err).To(HaveOccurred())
+		})
+	})
+
+	Context("When preparing a new input builder", func() {
+		Context("with a nil CertProject value", func() {
+			It("should return an error", func() {
+				_, err := NewCertificationInput(nil)
+				Expect(err).To(HaveOccurred())
+			})
+		})
+
+		Context("with a non-nil CertProject value as input", func() {
+			p := &CertProject{
+				ID:                  "myID",
+				CertificationStatus: "Pending",
+				Container: Container{
+					DockerConfigJSON: "",
+					Type:             "",
+					ISVPID:           "",
+					Registry:         "",
+					Repository:       "",
+					OsContentType:    "",
+					Privileged:       false,
+				},
+				Name:          "My Project",
+				ProjectStatus: "Some Status",
+				Type:          "Container",
+			}
+
+			builder, err := NewCertificationInput(p)
+
+			It("should not return an error", func() {
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			It("should return a certification input builder with the project embedded", func() {
+				Expect(builder.certificationInput.CertProject.ID).To(Equal(p.ID))
+			})
+
+			It("should fail to finalize with no cert image", func() {
+				_, err := builder.Finalize()
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should fail to bind a certimage from a file that does not exist", func() {
+				f := filepath.Join(tmpdir, "does-not-exist")
+				builder.WithCertImageFromFile(f)
+				Expect(builder.CertImage).To(BeNil())
+			})
+
+			It("should allow binding a valid cert image read from a file", func() {
+				f := filepath.Join(tmpdir, "certimage")
+				certimg := CertImage{
+					ID:                     "foo",
+					Certified:              false,
+					Deleted:                false,
+					DockerImageDigest:      "",
+					DockerImageID:          "",
+					ImageID:                "",
+					ISVPID:                 "",
+					ParsedData:             &ParsedData{},
+					Architecture:           "",
+					RawConfig:              "",
+					Repositories:           []Repository{},
+					SumLayerSizeBytes:      0,
+					UncompressedTopLayerId: "",
+				}
+				bts, err := json.Marshal(certimg)
+				Expect(err).ToNot(HaveOccurred())
+				os.WriteFile(f, bts, 0o0755)
+
+				err = builder.storeCertImage(f)
+				Expect(err).ToNot(HaveOccurred())
+
+				Expect(builder.CertImage.ID).To(Equal(certimg.ID))
+			})
+
+			It("should not bind an invalid cert image from file", func() {
+				f := filepath.Join(tmpdir, "certimage.invalid ")
+				err := os.WriteFile(f, []byte("\tHello world!\n"), 0o0755)
+				Expect(err).ToNot(HaveOccurred())
+
+				err = builder.storeCertImage(f)
+				Expect(err).To(HaveOccurred())
+			})
+
+			It("should fail to finalize with no test results", func() {
+				_, err := builder.Finalize()
+				Expect(err).To(HaveOccurred())
```

Comment:
Are you assuming that these are being run in any particular order? Doesn't ginkgo randomize tests?

Reply:
Looks like ginkgo randomizes the top-level specs, but everything within a container will be run in order. 

https://onsi.github.io/ginkgo/#spec-randomization

That said, randomize all is a recommended practice. I'm not sure if that breaks our testing today. I'm guessing wrapping this in before/after each would be the way to go if we want to prepare for --randomize-all? Bah.

Reply:
I'm ok within a spec to rely on it, I just wasn't sure what the default behaviour was.